Validation: Don't fail if the bundle path is same as default path

praveenkumar opened the pull request:
ValidateBundle function try to get the bundle info if the bundle is
already extracted or run ValidateBundlePath which in case check if the
bundle exist for provided path and for default bundle it give error in
case not persent in the default location. Ideally ValidateBundle should
just return nil in case default bundle path is provided.

With this change, same function can consme as part of `cmd/setup` which
right now have addition logic to ignore default bundle path.

praveenkumar committed:
Validation: Don't fail if the bundle path is same as default path

ValidateBundle function try to get the bundle info if the bundle is
already extracted or run ValidateBundlePath which in case check if the
bundle exist for provided path and for default bundle it give error in
case not persent in the default location. Ideally ValidateBundle should
just return nil in case default bundle path is provided.

With this change, same function can consme as part of `cmd/setup` which
right now have addition logic to ignore default bundle path.

praveenkumar committed:
cmd/setup: Use ValidateBundle for pre validation checks

In previous commit 5fbcac2477acfd87557e77c58619955e37ed303a ,
ValidateBundle now handle the default bundlepath also so we shouldn't
have a different logic here.

praveenkumar committed:
Preflight: Remove ValidateBundle logic as part of bundleExtract

This logic we have in `fixBundleExtract` function is kind of unnecessary
because ValidationBundle already part of setup,start and config set
command so having same logic doesn't make sense.

cfergeau commented on pkg/crc/preflight/preflight_checks_common.go:
```diff
@@ -110,17 +110,6 @@ func fixBundleExtracted(bundlePath string, preset crcpreset.Preset) func() error
-			var e *validation.InvalidPath
-			if !errors.As(err, &e) {
-				return err
-			}
```
This may be slightly better than what ValidateBundlePath is doing though, this does not ignore all errors, just invalid paths.

praveenkumar replied:
But ignoring the invalid path at this step doesn't make sense, this function suppose to fix the bundle extraction and all the errors are handled even invalid path one as part of validation run of different command.

cfergeau commented on pkg/crc/preflight/preflight_checks_common.go:
```diff
@@ -110,17 +110,6 @@ func fixBundleExtracted(bundlePath string, preset crcpreset.Preset) func() error
-			}
-			if bundlePath != constants.GetDefaultBundlePath(preset) {
-				/* This message needs to be improved when the bundle has been set in crc config for example */
-				return fmt.Errorf("%s is invalid or missing, run 'crc setup' to download the bundle", bundlePath)
```
And this error message seems better than what we get in ValidateBundlePath?

gbraad replied:
not really following what you try to say 🙄

praveenkumar replied:
Shouldn't we have this as soon as `bundle` argument is used and tell user the provided path is invalid or missing so either user run it with correct path or or just don't provide any bundle path at all.

cfergeau replied:
> And this error message seems better than what we get in ValidateBundlePath?

This removes a relatively clear error message (even if I'm not sure it's ever shown to the user): `fmt.Errorf("%s is invalid or missing, run 'crc setup' to download the bundle", bundlePath)`
This is more useful than:
```
$ ./out/linux-amd64/crc setup --bundle=/foo/bar
file '/foo/bar' does not exist
```

praveenkumar replied:
If that is just related to error message then we can update it to validation function and this error message is not even shown to user (Atleast I didn't find a hack which can use this code flow).

Error can be updated to following as part of validation?
```
fmt.Errorf("%s is invalid or missing, run 'crc setup' to download the bundle", bundlePath)
```

cfergeau replied:
I'd say "to download the default bundle"

praveenkumar replied:
@cfergeau updated

cfergeau reviewed (commented):
Overall looks good, a few questions though.

praveenkumar commented:
/retest

praveenkumar committed:
validation: Change InvalidPath => invalidPath (make it private)

invalidPath type only used in the validation so no point to make it
public.

praveenkumar committed:
Validation: Put more meaningful message in case invalid bundle path

With this PR now end user will observe more meaningful message to run
`crc setup` in case want to use the default bundle.
```
$ ./crc config set bundle /foo/bar
Value '/foo/bar' for configuration property 'bundle' is invalid, reason: /foo/bar is invalid or missing, run 'crc setup' to download the default bundle

$ ./crc start -b /foo/bar
/foo/bar is invalid or missing, run 'crc setup' to download the default bundle
```
